### PR TITLE
Implemented product availability get and update

### LINKED ISSUE #38: Feat: Product Availability
### Description

As a vendor, I should be able to mark a product as available or unavailable manually, so that buyers can view and purchase the available products. Products may need to be marked as unavailable due to reasons such as expiry date or out-of-stock situations.

### Acceptance Criteria

- [x] Implement an API endpoint for updating product availability
- [x] Validate the requesting user's role as a vendor
- [x] Validate the request payload (product ID, availability status)
- [x] Retrieve the product from the database based on the provided product ID
- [x] Ensure the product belongs to the requesting vendor
- [x] Update the product availability status in the database
- [x] Implement error handling and appropriate response messages for validation failures
- [x] Implement input sanitization

### Developer Notes

- Document the API endpoint with Swagger
- Define the request payload structure for updating product availability
- Implement necessary validation checks for the request payload
- Update the product availability status in the database
- Handle scenarios where the product does not exist or does not belong to the requesting vendor
- Write unit tests to cover different scenarios and edge cases

### Related Issues

- #37 

### Definition of Done

- API endpoint for updating product availability is implemented and functional
- Validation checks are implemented for the requesting user's role and request payload
- Product availability status is updated in the database
- Appropriate error handling and response messages are implemented
- Input sanitization measures are in place
- Unit tests are written and passing
- Code adheres to coding standards and best practices
- API documentation with Swagger is completed

### PR DESCRIPTION
**Pull request by niyobern**
# Product Availability Feature

## Summary
This pull request implements a new feature that allows vendors to update and query the availability of their products. Additionally, the product availability is automatically set to false when the quantity reaches zero.

## Changes
- Vendors can now update the availability status of their products.
- Vendors can query the current availability status of their products.
- Added a function that updates product availability when called inside controllers that modify product quantity.

```typescript
const productQuantityWatch = async (product: Product) => {
    const quantity = product.quantity;

    if (quantity > 0) {
        product.isAvailable = true;
    } else {
        product.isAvailable = false;
    }

    await productRepository.save(product)

    return product;
};
```

## How to Test
1. Update the product availability using the new endpoint.
2. Query the product to verify the updated availability status.
3. Reduce the product quantity to zero and verify that the availability status is automatically set to false.
4.  Increase the product availability from zero and verify that the product availability is set to true.

## Impact
This feature enhances the vendor's ability to manage their products more effectively, providing real-time updates on product availability based on quantity.

## Swagger documentation

**You can check the documentation at path `/api-docs`**

- You need to authenticate as per the authentication/authorization flow defined in the openapi schema
- Access those new endpoints and to get or update availability of any product 
- The request body, path parameters and response body is documented

image

## Additional Information
Please ensure to run the updated tests included in this PR to verify that everything is working as expected.

### CODE REVIEW

**Review by EddyShimwa — DISMISSED** [review]
Looks Good to me!

**Comment by ambroisegithub** [comment]
Please Add the screenshot of your swagger documentation and testing passing!!!

**Comment by niyobern** [comment]
> Please Add the screenshot of your swagger documentation and testing passing!!!

@ambroisegithub Swagger documentation screenshot are now included but there is no point in ataching screenshots of tests passing while we have the full log in github actions [https://github.com/atlp-rwanda/dynamites-ecomm-be/actions/runs/9147905258/job/25149832973](https://github.com/atlp-rwanda/dynamites-ecomm-be/actions/runs/9147905258/job/25149832973)

**Comment by ambroisegithub** [comment]
Add testing screenshot of your local one!!!

**Comment by niyobern** [comment]
> Add testing screenshot of your local one!!!

Here it is but with nothing additional.

image

**Review by wayneleon1 — DISMISSED** [review]
Looks Good to me!

**Comment by ambroisegithub** [comment]
It looks super for me!!!

**Review by bertrandshema — CHANGES_REQUESTED** [review]
Why do we have files for one just features, if we do this we will have an app with many files. since product availability is about products, please add them in product.test.ts, productcontroller.ts, product.routes.ts and product.doc.ts

**Comment by niyobern** [comment]
> Why do we have files for one just features, if we do this we will have an app with many files. since product availability is about products, please add them in product.test.ts, productcontroller.ts, product.routes.ts and product.doc.ts

Removed my controller, test, router and doc files

**Review by jkarenzi — DISMISSED** [review]

File: src/routes/index.ts
Comment (wayneleon1) [inline src/routes/index.ts:1]: can you remove all those commented lines if there are no long  necessary

**Review by wayneleon1 — CHANGES_REQUESTED** [review]

**Review by wayneleon1 — DISMISSED** [review]

**Review by bertrandshema — COMMENTED** [review]
It is good now.

**Review by bertrandshema — DISMISSED** [review]
It is good now

**Review by Dawaic6 — APPROVED** [review]

**Review by wayneleon1 — APPROVED** [review]
Looks good

**Review by dushimeemma — APPROVED** [review]
Good job 💪

**Pull request merged**